compiler/src/dotty/tools/repl/ScriptEngine.scala

```diff
@@ -0,0 +1,90 @@
+package dotty.tools
+package repl
+
+import java.io.{Reader, StringWriter};
+import javax.script.{AbstractScriptEngine, Bindings, ScriptContext, ScriptEngine => JScriptEngine, ScriptEngineFactory, ScriptException, SimpleBindings}
+
+/** A JSR 223 (Scripting API) compatible wrapper around the REPL for improved
+ *  interoperability with software that supports it.
+ *
+ *  It works by instantiating a new script engine through the script engine manager.
+ *  The script engine provides a eval method to evaluate scripts in string form.
+ *  Example use:
+ *
+ *  val m = new javax.script.ScriptEngineManager()
+ *  val e = m.getEngineByName("scala")
+ *  println(e.eval("42"))
+ *
+ *  @param factory the script engine factory
+ */
+class ScriptEngine(factory: ScriptEngine.Factory) extends AbstractScriptEngine {
+  val driver = factory.driver
+  val rendering = new Rendering
+  var state: State = driver.initialState
+
+  def getFactory(): ScriptEngineFactory = factory
+
+  def createBindings: Bindings = new SimpleBindings
+
+  /* Evaluate with the given context. */
+  @throws[ScriptException]
+  def eval(script: String, context: ScriptContext): Object = {
+    val vid = state.valIndex
+    state = driver.run(script)(state)
+    val oid = state.objectIndex
+    Class.forName(s"rs$$line$$$oid", true, rendering.classLoader()(state.context))
```

Comment:
Hardcoding implementation details like the weird classname here isn't great. I haven't looked in details at what this does but could it be achieved by some addition to the API of the REPL instead ?

Reply:
Yes. The class name is computed in Rendering.valueOf(symbol) as called from ReplDriver.displayDefinitions/Members(), themselves called from ReplDriver.compile(). In the ScriptEngine, I am calling ReplDriver.run() which calls interpret() which calls compile(). So in theory, one can decompose these calls to get access to the CompilationUnit and get the Symbol from the tpd.Tree and compute the newestWrapper to get the name.

Reply:
Or maybe State could have an additional field for the symbol of the last wrapper module ?